src/main/java/org/broadinstitute/hellbender/utils/runtime/StreamingProcessController.java

```diff
@@ -0,0 +1,598 @@
+     * @param <T> Type of items to be written to the stream.
+     * @return {@link AsynchronousStreamWriterService}
+     */
+    public <T> AsynchronousStreamWriterService<T> getAsynchronousStreamWriterService(final OutputStream outputStream) {
+        Utils.nonNull(outputStream);
+        return new AsynchronousStreamWriterService<>(executorService, outputStream);
+    }
+
+    /**
+     * Close the FIFO; called on controller termination
+     */
+    private void closeFIFO() {
+        if (fifoFile != null) {
+            fifoFile.delete();
+            fifoTempDir.delete();
+        }
+    }
+
+    /**
+     * Return true if either stdout or stderr ends with a synchronization string
+     */
+    private boolean scanForSynchronizationPoint(final ProcessOutput processOutput, final String synchronizationString) {
+        final StreamOutput stdOut = processOutput.getStdout();
+        if (stdOut != null) {
+            final String output = stdOut.getBufferString();
+            if (output != null && output.endsWith(synchronizationString)){
+                return true;
+            }
+        }
+        final StreamOutput stdErr = processOutput.getStderr();
+        if (stdErr != null) {
+            final String output = stdErr.getBufferString();
+            if (output != null && output.endsWith(synchronizationString)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Submit a task to start listening for output
+     */
+    private void startListeners() {
+        // Submit runnable task to start capturing.
+        if (stdOutFuture == null) {
+            stdOutFuture = executorService.submit(
+                    new OutputCapture(
+                            new CapturedStreamOutputSnapshot(settings.getStdoutSettings(), process.getInputStream(), System.out),
+                            ProcessStream.Stdout,
+                            this.getClass().getSimpleName(),
+                            controllerId));
+        }
+        if (!settings.isRedirectErrorStream() && stdErrFuture == null) {
+            // don't waste a callable on stderr if its just going to be redirected to stdout
+            stdErrFuture = executorService.submit(
+                    new OutputCapture(
+                            new CapturedStreamOutputSnapshot(settings.getStderrSettings(), process.getErrorStream(), System.err),
+                            ProcessStream.Stderr,
+                            this.getClass().getSimpleName(),
+                            controllerId));
+        }
+    }
+
+    /**
+     * Stops the process from running and tries to ensure the process is cleaned up properly.
+     * NOTE: sub-processes started by process may be zombied with their parents set to pid 1.
+     * NOTE: capture threads may block on read.
+     */
+    @Override
+    protected void tryDestroy() {
+        if (stdErrFuture != null && !stdErrFuture.isDone()) {
```

Comment:
It doesn't look like this is actually killing the process anywhere?  Just closing the various outputs?